### PR TITLE
Revent & Misc Fixes

### PR DESCRIPTION
**Pull request by marcbonnici**

### CODE REVIEW

File: wa/framework/getters.py [hunk 1]
```diff
@@ -80,7 +80,12 @@ def get_from_location(basepath, resource):
         path = os.path.join(basepath, 'bin', resource.abi, resource.filename)
         if os.path.exists(path):
             return path
-    elif resource.kind in ['apk', 'jar', 'revent']:
+    elif resource.kind == 'revent':
+        path = os.path.join(basepath, 'revent_files')
+        if os.path.exists(path):
```
Comment (setrofim) [inline wa/framework/getters.py]: This needs to check in both locations. if the file is not found in "revent_files" subdirectory, look in the base resource directory.

File: wa/workloads/angrybirds_rio/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+class AngryBirdsRio(ReventWorkload):
+
+    name = 'angrybirds_rio'
+    description = """
+    Angry Birds Rio game.
+
+    The sequel to the very popular Android 2D game.
+    """
+    package = 'com.rovio.angrybirdsrio'
+    activity = 'com.rovio.ka3d.App'
```
Comment (setrofim) [inline wa/workloads/angrybirds_rio/__init__.py]: Are these still necessary? IIRC, the new APK workload should auto-discover package and activity from APK info.

File: wa/framework/target/descriptor.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def get_descriptions(self):
 
 
 COMMON_TARGET_PARAMS = [
-    Parameter('working_directory', kind=str,
+    Parameter('working_directory', kind=str, default='/sdcard/wa-working',
```
Comment (setrofim) [inline wa/framework/target/descriptor.py]: This default is android-specific; the Parameter isn't.

File: wa/framework/workload.py [hunk 4]
```diff
@@ -214,50 +255,60 @@ def __init__(self, workload, target, setup_timeout=5 * 60, run_timeout=10 * 60):
...
-        command = '{} replay {}'.format(self.on_target_revent_binary, self.on_target_run_revent)
-        self.logger.debug('Replaying {}'.format(os.path.basename(self.on_target_run_revent)))
-        self.target.execute(command, timeout=self.run_timeout)
+    def init_resources(self, resolver):
+        self.revent_setup_file = resolver.get(ReventFile(owner=self.workload,
+                                                         stage='setup',
+                                                         target=self.target.model))
+        self.revent_run_file = resolver.get(ReventFile(owner=self.workload,
+                                                       stage='run',
+                                                       target=self.target.model))
```
Comment (setrofim) [inline wa/framework/workload.py]: The `setup` recording should be optional. There should also be optional recording for `extract_results` and `teardown` stages.

**Comment by setrofim** [comment]
The Angry Birds Rio workload (which subclasses `ReventWorkload`) commit occurs before the commit that adds `ReventWorkload`

File: wa/utils/revent.py [hunk 5]
```diff
@@ -16,9 +16,11 @@
 from __future__ import division
```
Comment (setrofim) [inline wa/utils/revent.py]: Incorrect capitalization/spelling of `ReventRecorder` in the commit message

File: wa/commands/record.py [hunk 6]
```diff
@@ -173,6 +173,57 @@ def workload_record(self, args):
...
+    name = 'replay'
+    description = '''
+    Replay a revent recording
+
+    Revent allows you to record raw inputs such as screen swipes or button presses.
+    See ``wa show record`` to see how to make an revent recording.
+    '''
+
+    def initialize(self, context):
+        self.parser.add_argument('revent', help='The name of the file to replay')
```
Comment (setrofim) [inline wa/commands/record.py]: Rename to "recording". Set metavar to "FILE"

File: wa/commands/record.py [hunk 7]
```diff
@@ -49,46 +50,132 @@ class RecordCommand(Command):
...
+        self.logger.info('Deploying {}'.format(args.workload))
+        workload = pluginloader.get_workload(args.workload, self.target)
+        workload.apk.init_resources(context.resolver)
+        workload.apk.setup(context.resolver)
+        sleep(workload.loading_time)
+
+        output_path = args.output or os.path.join(workload.dependencies_directory,
+                                                  'revent_files')
+        self.record(setup_file, 'SETUP', output_path)
+        self.record(run_file, 'RUN', output_path)
```
Comment (setrofim) [inline wa/commands/record.py]: `setup` recording should be optional. There should also be optional stages for `extract_results` and `teardown`.

The user should be able to specify with parameters which stages, in addition to `run`, they wish to record. E.g. -s for setup, -e for extract, -t for teardown, -a for all.

File: wa/commands/record.py [hunk 8]
```diff
@@ -49,46 +50,132 @@ class RecordCommand(Command):
...
     def initialize(self, context):
         self.parser.add_argument('-d', '--device', metavar='DEVICE',
                                  help='''
                                  Specify the device on which to run. This will
                                  take precedence over the device (if any)
                                  specified in configuration.
                                  ''')
-
-    def execute(state, args):
+        self.parser.add_argument('-o', '--output', help='Directory to save the recording in')
```
Comment (setrofim) [inline wa/commands/record.py]: This should a file rather than directory. 
Set metavar to "FILE"

**Comment by marcbonnici** [comment]
Pull request changes implemented.

**Comment by marcbonnici** [comment]
ApkUIWorkload base class added.

**Comment by marcbonnici** [comment]
Decorators Updated

**Pull request merged**